### ISSUE TITLE
py2py3: strings vs bytes vs unicode in CRABServer REST 

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6388]
when building with WMCore  1.4.6.pre2 we discovered that new code for
REST params validation with py2py3 compatibility makes CRAB REST fail
see:
https://github.com/dmwm/CRABServer/issues/6373#issuecomment-774296480

**Comment by belforte** [comment]
repost here info from #6373 to keep a complete thread:


The immediate problem is that the WMCore/REST parameter validation was changed in
https://github.com/dmwm/WMCore/commit/d6b9591d0325b523bb63e5b693aa8aab75234e2d
so that check_str now returns a list of bytes
https://github.com/dmwm/WMCore/blob/a4f60f75d5cfea05e4d3<f06d4e02f1acd23ba01a/src/python/WMCore/REST/Validation.py#L31

later inside CRABServer code such paramenter cominng from a user submisssion HTTP PUT is compared with e.g. a list of sites retrieved from CRIC (JSON, i.e. a list of unicode strigs) and while  `'a' is in [u'a']` is True, after the conversion of `'a'`to  `future.types.newbytes.newbytes` , that is False.

I fully agree that strings and bytes and unicode are different things and that the current code is just "sloppy".
But now that the careless handling of unicode in CRABServer has surfaced, I am afraid it will take some time and much more effort to make things work.

Almost only for confirmation of this interpretation I have managed to submit my test task through the CRABServer on cmsweb-test2, built with WMCore 1.4.6.pre2 changing this
https://github.com/dmwm/CRABServer/blob/35d6d99973607a4a12730a00db92f0f84aeb6355/src/python/CRABInterface/RESTUserWorkflow.py#L351
into
```python
 validate_ustrlist("sitewhitelist", param, safe, RX_CMSSITE) 
```
and similarly in other two places.


FWIW I decided to give it a try with
```
for f in `ls *py`; do sed -i -e 's/validate_str/validate_ustr/g' $f; done
```
I issued it a couple of times because some line has >1 occurrences of validate_str 

and restarted crab-dev-tw01.

With that crab client and TW work, but there's an error when PostJob fails to update status of transferred files
```
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa ENTER cmsweb_analysis_dev@int2r CRABInterface.RESTBaseAPI.RESTBaseAPI (POST dev filetransfer
s) inuse=0 idle=1
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa (previously nNXxVzhvLYHs) connected, client: 11.2.0.4.0, server: 11.2.0.4.0, stmtcache: 50
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa ping
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa check [select sysdate from dual]
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa connection established
[05/Feb/2021:23:28:31]  SERVER DATABASE ERROR 500/403 Execution error __builtins__.TypeError 750a966bb8be0c19dece3081c1af9dbe [instance: dev] (translate() takes exactly one argument (2 given)); last statement: None; binds: None, None; offset: None
[05/Feb/2021:23:28:31]    Traceback (most recent call last):
[05/Feb/2021:23:28:31]      File "/data/srv/HG2103a-comp/sw.ddirmait/slc7_amd64_gcc630/cms/crabserver/v3.210204/lib/python2.7/site-packages/WMCore/REST/Server.py", line 1772, in dbapi_wrapper
[05/Feb/2021:23:28:31]        return handler(*xargs, **xkwargs)
[05/Feb/2021:23:28:31]      File "/data/srv/HG2103a-comp/sw.ddirmait/slc7_amd64_gcc630/cms/crabserver/v3.210204/lib/python2.7/site-packages/CRABInterface/RESTFileTransfers.py", line 167, in post
[05/Feb/2021:23:28:31]        ids = (kwargs['list_of_ids'].translate(None,"[ ]'")).split(",")
[05/Feb/2021:23:28:31]    TypeError: translate() takes exactly one argument (2 given)
[05/Feb/2021:23:28:31] crabserver-7b49f7f484-hv9zk 137.138.148.19 "POST /crabserver/dev/filetransfers HTTP/1.1" 500 Internal Server Error [data: 3119 in 745 out 11720 us ] [auth: OK "/DC=org/DC=terena/DC=tcs/C=IT/O=Istituto Nazionale di Fisica Nucleare/CN=Stefano Belforte belforte@infn.it" "" ] [ref: "" "CRABServer/0.0.0" ]
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa disconnecting
[05/Feb/2021:23:28:31]  RESTSQL:vzrkOsniCXFa DISCONNECTED cmsweb_analysis_dev@int2r timeout=300 inuse=0 idle=0
```

**Comment by belforte** [comment]
And of course there can be many other places, besides REST call handling, where CRAB code suffers of string/unicode confusion. All in all  this code started in a time when people called "json" the string conversion of a python dictionary, and I have been fighting unicode confusion for years. Simply my cure usually was to cast everything to `str`. But with these `newstrings` around which are not really list of bytes anymore... more changes and more type safety may be needed.

**Comment by belforte** [comment]
the problem above 
```
[05/Feb/2021:23:28:31]      File "/data/srv/HG2103a-comp/sw.ddirmait/slc7_amd64_gcc630/cms/crabserver/v3.210204/lib/python2.7/site-packages/CRABInterface/RESTFileTransfers.py", line 167, in post
[05/Feb/2021:23:28:31]        ids = (kwargs['list_of_ids'].translate(None,"[ ]'")).split(",")
[05/Feb/2021:23:28:31]    TypeError: translate() takes exactly one argument (2 given)
```
is another unicode vs. string mess.

PostJob makes a POST call with this param
```
{'list_of_ids': [u'8a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1'], 'list_of_transfer_state': ['NEW'], 'asoworker': 'schedd', 'subresource': 'updateTransfers'}
```
this is first urlenconde, then passed to pycurl
```
138  ->	                response, datares = self['conn'].request(url, data, encode=True, headers=headers, verb=verb, doseq=True,
139  	                                                         ckey=self['key'], cert=self['cert'], capath=caCertPath,
140  	                                                         verbose=self['verbose'])
141  	            except Exception as ex:
142  	                #add here other temporary errors we need to retry
143  	                if (not retriableError(ex)) or (i == self['retry']):
(Pdb) url
'https://cmsweb-test2.cern.ch:8443/crabserver/dev/filetransfers'
(Pdb) data
'list_of_ids=%5B%278a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1%27%5D&list_of_transfer_state=%5B%27NEW%27%5D&asoworker=schedd&subresource=updateTransfers'
(Pdb) 
```
inside`pycurl_manager`:
```
(Pdb) n
> /data/srv/glidecondor/condor_local/spool/8883/0/cluster60918883.proc0.subproc0/WMCore/Services/pycurl_manager.py(249)request()
-> curl = pycurl.Curl()
(Pdb) l
244  	
245  	    def request(self, url, params, headers=None, verb='GET',
246  	                verbose=0, ckey=None, cert=None, capath=None,
247  	                doseq=True, encode=False, decode=False, cainfo=None, cookie=None):
248  	        """Fetch data for given set of parameters"""
249  ->	        curl = pycurl.Curl()
250  	        bbuf, hbuf = self.set_opts(curl, url, params, headers, ckey, cert, capath,
251  	                                   verbose, verb, doseq, encode, cainfo, cookie)
252  	        curl.perform()
253  	        if verbose:
254  	            print(verb, url, params, headers)
(Pdb) url
'https://cmsweb-test2.cern.ch:8443/crabserver/dev/filetransfers'
(Pdb) params
'list_of_ids=%5B%278a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1%27%5D&list_of_transfer_state=%5B%27NEW%27%5D&asoworker=schedd&subresource=updateTransfers'
(Pdb) verb
'POST'
(Pdb) 
```
and when this is sent to crabserver REST in next line `curl.perform()` it is received by CRAB RESTFileTransfers which inherits from WMCore/REST/Server/RESTEntity and here the validation method receives
```
{'list_of_ids': u"['8a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1']", 'list_of_transfer_state': u"['NEW']", 'asoworker': u'schedd'}
```
since I have now set all validation to use `validate ustr`, (note the `list_of_ids` is handled like a plain string, the output of the validation is
```
(Pdb) safe
RESTArgs(args=[], kwargs={'username': None, 'list_of_publication_state': None, 'list_of_transfer_state': u"['NEW']", 'subresource': u'updateTransfers', 'list_of_ids': u"['8a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1']", 'asoworker': u'schedd', 'list_of_fts_id': None, 'list_of_failure_reason': None, 'publish_flag': None, 'list_of_fts_instance': None, 'list_of_retry_value': None, 'time_to': None})
```
Note
```
list_of_ids': u"['8a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1']"
```
later in the code... this line which is mean to transform a " a string printout of a list, into a list object"
```
ids = (kwargs['list_of_ids'].translate(None,"[ ]'")).split(",")
```
is presented with
```
(Pdb) kwargs['list_of_ids']
u"['8a467dec017620a2e5bb702baf66d5fc161597cc86c6bcd221af5ca1']"
```
which rightfully fails because the python `str.translate` method does not accept the list of chars to delete in case of unicode https://docs.python.org/2.7/library/stdtypes.html?highlight=translate#str.translate

I suspect that we need to replace the use of `translate` here and in many other places with something more solid.

**Comment by belforte** [comment]
at least in this case the very stupid change from
https://github.com/dmwm/CRABServer/blob/55ad79cd32d365177a97394737145e7ab5aaf9d8/src/python/CRABInterface/RESTFileTransfers.py#L167
to
```
         ids = (str(kwargs['list_of_ids']).translate(None,"[ ]'")).split(",")
```
has apparently solved.

I suspect that at least this will not be py3 compliant. As to working in all possible use case we have now... I don't know. And inserting str() in all proper places is quite some work.

**Comment by belforte** [comment]
promising:
```
from Utilis.Utilities import makeList
ids = makeList(kwargs['list_of_ids'])
```

**Comment by belforte** [comment]
did the change from translate to makeList, resatarted crab-dev-tw01, sumitted a 4 job task, all jobs and PostJobs completed OK. 
```
Task name:			210207_221654:belforte_crab_20210207_231647
Grid scheduler - Task Worker:	crab3@vocms0194.cern.ch - crab-dev-tw01
Status on the CRAB server:	SUBMITTED
Task URL to use for HELP:	https://cmsweb-test2.cern.ch/crabserver/ui/task/210207_221654%3Abelforte_crab_20210207_231647
Dashboard monitoring URL:	https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&var-user=belforte&var-task=210207_221654%3Abelforte_crab_20210207_231647&from=1612732614000&to=now
In case of issues with the dashboard, please provide feedback to hn-cms-computing-tools@cern.ch
Status on the scheduler:	COMPLETED

Jobs status:                    finished     		100.0% (4/4)

Publication status of 1 dataset(s):	new          		100.0% (4/4)

```

I think we are not running a Publisher for the dev instance. will try that tomorrow.
In the meanwhile I sent a 200 job task, and a 600job one just to get some "variety".

**Comment by belforte** [comment]
so far so good but jobs never failed so retries from PJ have not been exercised. Sending more.

**Comment by belforte** [comment]
I pointed Publisher_schedd in crab-preprod-tw01 to REST instance `test2` and all my pending publications have been done.
More tests needed. May we be out of the woods so easily ?
I'l prepare a PR with changes which I have done in `test2`.

**Comment by belforte** [comment]
this Publisher works for my tasks but is failing on some old ones from ddirmait from January. Need to check REST logs again
```
021-02-08 09:14:54,308:ERROR: Error updating status for DocumentId: c0dbd19889bcef52f6b80149a216511f386decbddc83f646a3020cb7 lfn: /store/temp/user/ddirmait.72572fa442852e2527a58b84eb3acd90c4af6dd0/GenericTTbar/CRAB3_tutorial_May2015_MC_analysis/210131_101528/0000/output_14.root
2021-02-08 09:14:54,308:ERROR: Error reason: url=https://cmsweb-test2.cern.ch:8443/crabserver/dev/filetransfers, code=500, reason=Internal Server Error, headers={'Content-Length': '745', 'X-Error-Http': '500', 'X-Rest-Time': '65007.925 us', 'Server': 'CherryPy/17.4.0', 'Connection': 'close', 'X-Error-Detail': 'Execution error', 'X-Rest-Status': '403', 'Date': 'Mon, 08 Feb 2021 08:14:54 GMT', 'Content-Type': 'text/html;charset=utf-8', 'CMS-Server-Time': 'D=135682 t=1612772094168370', 'X-Error-Id': '5a290bd549e2385ee905f95a485c39c2'}
2021-02-08 09:16:55,188:ERROR: Fatal error trying to connect to https://cmsweb-test2.cern.ch:8443/crabserver/dev/filetransfers using list_of_publication_state=DONE&subresource=updatePublication&list_of_ids=a182b7e59a57f81d4b57fba9c0b48fce36536710c9a4e312fe5deacd&asoworker=schedd&list_of_failure_reason=&list_of_retry_value=1
2021-02-08 09:16:55,189:ERROR: Error updating status for DocumentId: a182b7e59a57f81d4b57fba9c0b48fce36536710c9a4e312fe5deacd lfn: /store/temp/user/ddirmait.72572fa442852e2527a58b84eb3acd90c4af6dd0/GenericTTbar/CRAB3_tutorial_May2015_MC_analysis/210131_101528/0000/output_10.root
```

**Comment by belforte** [comment]
somehow cmsweb-test2 was running old RESTFileTransfer.py again. Working on it

**Comment by belforte** [comment]
I suspect that this line messes up importing WMCore Utils:
https://github.com/dmwm/CRABServer/blob/55ad79cd32d365177a97394737145e7ab5aaf9d8/src/python/CRABInterface/RESTBaseAPI.py#L13

**Comment by belforte** [comment]
New error
```
[08/Feb/2021:10:38:21]      File "/data/srv/HG2103a-comp/sw.ddirmait/slc7_amd64_gcc630/cms/crabserver/v3.210204/lib/python2.7/site-packages/CRABInterface/RESTFileTransfers.py", line 214, in post
[08/Feb/2021:10:38:21]        binds['fail_reason'] = [reasons[num]]
[08/Feb/2021:10:38:21]    IndexError: list index out of range
```

**Comment by belforte** [comment]
this is the arg of the POST
```
list_of_publication_state=DONE&subresource=updatePublication&list_of_ids=45f2e54bd63bfcc2d33a606cfdadae8cbead61da4fbe7b2a76c39b60&asoworker=schedd&list_of_failure_reason=&list_of_retry_value=1
```
which indeed is not great viz. `list_of_failure_reason=` with nothing in RHS of `=`, but somehow used to work.

**Comment by belforte** [comment]
so the problem is somehow WMCore's `makeList` which when passed an empty string `u''` returns an empty list `[]` instead of a list containing the empty string `['']`
while I can look into changing the arg sent to the REST, this may (or may be not) interpreted as a defect in `makeList`.
Surely there's room for different opinions on what it should do in this case. Previous CRAB's `translate` trick would always return a list thanks to the `.split()` at the end, but it does not work on unicode.
@amaltaro does current behavior matches your expectation ?

**Comment by amaltaro** [comment]
> so the problem is somehow WMCore's makeList which when passed an empty string u'' returns an empty list [] instead of a list containing the empty string ['']

that is indeed the expected behaviour, see:
https://github.com/dmwm/WMCore/blob/ed40d33069bdddcd98ed5b8430d5ca6662e5941f/src/python/Utils/Utilities.py#L45-L46

Are you saying that now it returns a list with an empty string - `['']` - and that's what causes a problem here?

**Comment by belforte** [comment]
recap of all changes I did while debugging and which I need now to properly insert in GH:
- [ ] change validate_str to validate_ustr
- [ ] change TaskPublish to always send lists  args named `list_of...`
- [ ] rename CRABInterface/Utils to CRABInterface/Utilities
- [ ] change all things like `ids = (str(kwargs['list_of_ids']).translate(None,"[ ]'")).split(",")` to use WMCore/Utils/Utilities/makeList

**Comment by belforte** [comment]
> > so the problem is somehow WMCore's makeList which when passed an empty string u'' returns an empty list [] instead of a list containing the empty string ['']
> 
> that is indeed the expected behaviour, see:
> https://github.com/dmwm/WMCore/blob/ed40d33069bdddcd98ed5b8430d5ca6662e5941f/src/python/Utils/Utilities.py#L45-L46
thanks Alan, why did I not notice those two lines ???? oh well..
> Are you saying that now it returns a list with an empty string - `['']` - and that's what causes a problem here?
the other way around.
Now (I think like ever) it returns an empty list `[]` which is causing an error here. While in my case `['']` would have worked. I think there's room for having many opinions here. I only wanted to be sure that current behavior is not surprising you. I will of course take this into account when using `makeList`. Thanks

**Comment by belforte** [comment]
so far, so good:
```
belforte@lxplus720/TC3> crab status -d crab_20210208_150722/
CRAB project directory:		/afs/cern.ch/work/b/belforte/CRAB3/TC3/crab_20210208_150722
Task name:			210208_140729:belforte_crab_20210208_150722
Grid scheduler - Task Worker:	crab3@vocms0194.cern.ch - crab-dev-tw01
Status on the CRAB server:	SUBMITTED
Task URL to use for HELP:	https://cmsweb-test2.cern.ch/crabserver/ui/task/210208_140729%3Abelforte_crab_20210208_150722
Dashboard monitoring URL:	https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&var-user=belforte&var-task=210208_140729%3Abelforte_crab_20210208_150722&from=1612789649000&to=now
In case of issues with the dashboard, please provide feedback to hn-cms-computing-tools@cern.ch
Warning:			The following sites from the user site whitelist are blacklisted by the CRAB server: [u'T2_CH_CERN_HLT', u'T2_GR_Ioannina', u'T2_RU_ITEP', u'T2_RU_SINP', u'T2_IT_Pisa']. Since the CRAB server blacklist has precedence, these sites are not considered in the user whitelist.
Warning:			The following sites appear in both the user site blacklist and whitelist: [u'T2_CH_CERN']. Since the whitelist has precedence, these sites are not considered in the blacklist.
Status on the scheduler:	SUBMITTED

Jobs status:                    finished     		 77.8% (467/600)
				running      		  0.2% (  1/600)
				transferring 		 22.0% (132/600)

Publication status of 1 dataset(s):	acquired     		  2.2% ( 13/600)
(from CRAB internal bookkeeping in transferdb)
					done         		 50.0% (300/600)
					new          		 44.8% (269/600)
					unsubmitted  		  3.0% ( 18/600)


```

**Comment by belforte** [comment]
all problems listed above should be fixed in this PR https://github.com/dmwm/CRABServer/pull/6395
but I am keeping this open at least  until we make a build which runs a couple tasks through the whole chain

**Comment by belforte** [comment]
at the end my task above has 2 files whose publication is stuck in NEW.  Will investigate

**Comment by belforte** [comment]
At first sight looks like a PostJob bug, how is it possible that we never noticed ?
Looks like:
aso xfer failed, job was retried, second time ran at CERN (destination site) and there was no ASO, but transfer_state was left at 0 (NEW). Later Publisher looks for files with publication request flag AND xfer status DONE (do not publish files which we did not xfer).
Indeed @dciangot mumbled something at some point about retries+directstageout, but that was about retrying publications.. not this kind of stupid thing here.
I need to be well sure of course, before I put my hands in the PostJob.
Let's just say that atm there is no indication of other compatibility problems with WMCore 1.4+
We need to run more

**Comment by belforte** [comment]
a second submission of same task did not run at CERN and everything worked like a charm[*]
OTOH @ddaina found that submission w/o a sitewhitelist fails with an odd message
```
<....same as above....>
[08/Feb/2021:22:08:03]      File "/data/srv/HG2103a-comp/sw.ddirmait/slc7_amd64_gcc630/cms/crabserver/v3.210204/lib/python2.7/site-packages/CRABInterface/RESTUserWorkflow.py", line 361, in validate
[08/Feb/2021:22:08:03]        validate_ustrlist("TESTsitewhitelist", param, safe, RX_CMSSITE)
[08/Feb/2021:22:08:03]    KeyError: 'sitewhitelist'
```
I will debug this later today.

[*]
```
belforte@lxplus720/TC3> crab status -d crab_20210208_230724/
CRAB project directory:		/afs/cern.ch/work/b/belforte/CRAB3/TC3/crab_20210208_230724
Task name:			210208_220729:belforte_crab_20210208_230724
Grid scheduler - Task Worker:	crab3@vocms0197.cern.ch - crab-dev-tw01
Status on the CRAB server:	SUBMITTED
Task URL to use for HELP:	https://cmsweb-test2.cern.ch/crabserver/ui/task/210208_220729%3Abelforte_crab_20210208_230724
Dashboard monitoring URL:	https://monit-grafana.cern.ch/d/cmsTMDetail/cms-task-monitoring-task-view?orgId=11&var-user=belforte&var-task=210208_220729%3Abelforte_crab_20210208_230724&from=1612818449000&to=now
In case of issues with the dashboard, please provide feedback to hn-cms-computing-tools@cern.ch
Warning:			The following sites from the user site whitelist are blacklisted by the CRAB server: [u'T2_CH_CERN_HLT', u'T2_GR_Ioannina', u'T2_RU_ITEP', u'T2_RU_SINP', u'T2_IT_Pisa']. Since the CRAB server blacklist has precedence, these sites are not considered in the user whitelist.
Status on the scheduler:	COMPLETED

Jobs status:                    finished     		100.0% (600/600)

Publication status of 1 dataset(s):	done         		100.0% (600/600)
(from CRAB internal bookkeeping in transferdb)

Output dataset:			/GenericTTbar/belforte-Stefano-Test-2026-bb695911428445ed11a1006c9940df69/USER
Output dataset DAS URL:		https://cmsweb.cern.ch/das/request?input=%2FGenericTTbar%2Fbelforte-Stefano-Test-2026-bb695911428445ed11a1006c9940df69%2FUSER&instance=prod%2Fphys03

Warning: the max jobs runtime is less than 30% of the task requested value (60 min), please consider to request a lower value for failed jobs (allowed through crab resubmit) and/or improve the jobs splitting (e.g. config.Data.splitting = 'Automatic') in a new task.

Warning: the average jobs CPU efficiency is less than 50%, please consider to improve the jobs splitting (e.g. config.Data.splitting = 'Automatic') in a new task

Summary of run jobs:
 * Memory: 2MB min, 467MB max, 76MB ave
 * Runtime: 0:01:01 min, 0:12:39 max, 0:01:34 ave
 * CPU eff: 5% min, 71% max, 44% ave
 * Waste: 10:13:55 (39% of total)

Log file is /afs/cern.ch/work/b/belforte/CRAB3/TC3/crab_20210208_230724/crab.log
```

The PostJob problem from previous comment deserves its own issue. Let's leave it off this thread.

**Comment by belforte** [comment]
when I restarted Crabserver in debug mode to look at the problem with sitewhitelit https://github.com/dmwm/CRABServer/issues/6388#issuecomment-775761779 ... everything worked.
Mystery. Restarted in normal mode and keeps working.
Only change I knowingly did was to comment out a `print("IN VALIDATE")` line which I had inserted in RESTUserWorkflow.py at the beginning of the `validate` method the other day to help me debug the other problems.
I do not see why a print should result in memory corruption or similar, but surely the error was making no sense since it was happening in a submission in which neither siteblacklist nor sitewhitelist were defined and the two are checked in identical manner in the code, but only the second was failing !

Better this way.
No known problem at this point.
